### PR TITLE
improvement(Alternator): support creating tablets-enabled tables

### PR DESCRIPTION
**Pull request by yarongilor**
Tablets feature is currently supported by Alternator, but disabled by default (since LWT is not supported).
	It should be explicitly requested by the specified tag.
	fixes: #8197
	refs: scylladb/scylladb#18068
fixes: #8197
### Testing

- Test Alternator without LWT and TTL runs with tablets:
- - https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/13/
- - - [test](https://argus.scylladb.com/test/dfa0a7dd-68a5-4a6c-8d9c-a1131811fdbe/runs?additionalRuns[]=b18b5cd1-2346-4ab0-8b20-b2fe7382b16b) passed ok
- - - 
image

```
Aug 07 16:14:34.708495 alternator-no-ttl-1-loaders-no-lwt--db-node-b18b5cd1-1 scylla[6397]:  [shard 0:stmt] migration_manager - Create new Keyspace: KSMetaData{name=alternator_usertable_no_lwt, strategyClass=org.apache.cassandra.locator.NetworkTopologyStrategy, strategyOptions={"eu-west": "3"}, cfMetaData={}, durable_writes=true, tablets={"initial":0}, userTypes=org.apache.cassandra.config.UTMetaData@0x600008f8f1a8}

2024-08-07T16:14:35.158+00:00 alternator-no-ttl-1-loaders-no-lwt--db-node-b18b5cd1-1     !INFO | scylla[6397]:  [shard 0: gms] load_balancer - Node 15f609a5-151d-4f94-9a24-1fe6525aa8a3: rack=1a avg_load=1, tablets=4, shards=4, state=normal

```

### PR pre-checks (self review)

- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

### CODE REVIEW

**Comment by yarongilor** [comment]
Tested in: https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/10/

**Comment by yarongilor** [comment]
> Tested in: https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/12/
> 
> @fruch , @roydahan , the Alternator non-LWT [test](https://argus.scylladb.com/test/dfa0a7dd-68a5-4a6c-8d9c-a1131811fdbe/runs?additionalRuns%5B%5D=89ef35b8-79b5-4a50-99c3-6833ee8968bf) was appropriately configured with tablets, but then failed trying to configure Alternator TTL since it is not supported:
> 
> ```
> Aug  4 18:04:17 alternator-ttl-4-loaders-no-lwt-sis-db-node-89ef35b8-1 scylla[9348]:  [shard  0:stmt] migration_manager - Create new Keyspace: KSMetaData{name=alternator_usertable_no_lwt, strategyClass=org.apache.cassandra.locator.NetworkTopologyStrategy, strategyOptions={"eu-west": "3"}, cfMetaData={}, durable_writes=true, tablets={"initial":0}, userTypes=org.apache.cassandra.config.UTMetaData@0x6000055beea8}
> ```
> 
> ```
> botocore.exceptions.ClientError: An error occurred (ValidationException) when calling the UpdateTimeToLive operation: TTL not yet supported on a table using tablets (issue #16567). Create a table with the tag 'experimental:initial_tablets' set to 'none' to use vnodes.
> ```
> 
> The test yaml contains TTL so it is pre-configured in SCT by:
> 
> ```
>             if any('dynamodb.ttlKey' in str(cmd) for cmd in [prepare_cmd, stress_cmd]):
>                 self.alternator.update_table_ttl(node=node, table_name=alternator.consts.TABLE_NAME)
>                 self.alternator.update_table_ttl(node=node, table_name=alternator.consts.NO_LWT_TABLE_NAME)
> ```
> 
> we can consider modifying the condition of this PR to something like:
> 
> ```
>                 tablets_enabled = is_tablets_feature_enabled(session) and not any('dynamodb.ttlKey' in str(cmd) for cmd in [prepare_cmd, stress_cmd])
> ```
> 
> cc: @nyh

fixed by disabling tablets in case alternator TTL is in use:
```
-                tablets_enabled = is_tablets_feature_enabled(session)
+                tablets_enabled = is_tablets_feature_enabled(session)  # is tablets feature enabled in Scylla configuration.
+            prepare_cmd = self.params.get('prepare_write_cmd')
+            stress_cmd = self.params.get('stress_cmd')
+            is_ttl_in_workload = any('dynamodb.ttlKey' in str(cmd) for cmd in [prepare_cmd, stress_cmd])
+            if is_ttl_in_workload and SkipPerIssues(issues="scylladb/scylladb#16567", params=self.params):
+                tablets_enabled = False  # Alternator TTL is not yet supported with tablets.
             schema = self.params.get("dynamodb_primarykey_type")
             self.alternator.create_table(node=node, schema=schema)
             self.alternator.create_table(node=node, schema=schema, isolation=alternator.enums.WriteIsolation.FORBID_RMW,
                                          table_name=NO_LWT_TABLE_NAME, tablets_enabled=tablets_enabled)
-            prepare_cmd = self.params.get('prepare_write_cmd')
-            stress_cmd = self.params.get('stress_cmd')
-            if any('dynamodb.ttlKey' in str(cmd) for cmd in [prepare_cmd, stress_cmd]):
+
+            if is_ttl_in_workload:
```

File: sdcm/tester.py
Comment (fruch) [inline sdcm/tester.py:1127]: what this comment says ?

File: sdcm/tester.py
Comment (fruch) [inline sdcm/tester.py:1126]: I don't understand this, what's the point of setting it False ?

it's a bit confusing, since tablet might be enabled by default, but you don't want to enabled it, so piggy banking on `tablets_enabled` to extend the logic is a bit off.

`should_enable_tablets` and apply logic to it one by one, instead of mixing it, might be a bit more clear.

File: sdcm/tester.py
Comment (yarongilor) [reply sdcm/tester.py:1127]: rephrased: `Enable tablets for Alternator with LWT requires issue: "[Epic] tablets: support LWT #18068"`

File: sdcm/tester.py
Comment (yarongilor) [reply sdcm/tester.py:1126]: Modified to be clearer by:
```
-            if is_ttl_in_workload and SkipPerIssues(issues="scylladb/scylladb#16567", params=self.params):
-                tablets_enabled = False  # Alternator TTL is not yet supported with tablets (#16567).
-            # [Epic] tablets: support LWT #18068
+
+            # Enable tablets with Alternator TTL requires (#16567).
+            tablets_support_ttl = not SkipPerIssues(issues="scylladb/scylladb#16567", params=self.params)
+
+            # Enable tablets for Alternator with LWT requires issue: "[Epic] tablets: support LWT #18068"
             tablets_support_lwt = not SkipPerIssues(issues="scylladb/scylladb#18068", params=self.params)
+
+            tablets_supported_alternator_workload = not is_ttl_in_workload or tablets_support_ttl
             schema = self.params.get("dynamodb_primarykey_type")
             self.alternator.create_table(node=node, schema=schema,
-                                         tablets_enabled=tablets_enabled and tablets_support_lwt)
+                                         tablets_enabled=tablets_enabled and tablets_support_lwt and tablets_supported_alternator_workload)
             self.alternator.create_table(node=node, schema=schema, isolation=alternator.enums.WriteIsolation.FORBID_RMW,
-                                         table_name=NO_LWT_TABLE_NAME, tablets_enabled=tablets_enabled)
+                                         table_name=NO_LWT_TABLE_NAME,
+                                         tablets_enabled=tablets_enabled and tablets_supported_alternator_workload)
```

**Comment by yarongilor** [comment]
@pehala , @bhalevy , can you please advise? this PR is pending review/merge for 3 weeks.

**Comment by fruch** [comment]
> @pehala , @bhalevy , can you please advise? this PR is pending review/merge for 3 weeks.

It's better to first have review from the  relevant developers about such logic.

Once it is reviewed. would be merged

**Comment by bhalevy** [comment]
I don't know much about enabling tablets for alternator in sct (or even natively), so FWIW, without verifying the details, looks good to me

**Comment by pehala** [comment]
> > @pehala , @bhalevy , can you please advise? this PR is pending review/merge for 3 weeks.
> 
> It's better to first have review from the relevant developers about such logic.
> 
> Once it is reviewed. would be merged

@nyh Could you please have look, if logic in this PR is sound?

**Review by nyh — APPROVED** [review]
Looks correct logic to me (please run it as well to see it actually does what I think it does, but it does look correct to me).

**Review by fruch — APPROVED** [review]

**Comment by yarongilor** [comment]
@scylladb/qa-maintainers , please merge this, so we can rerun https://argus.scylladb.com/test/cc3a6940-a949-4765-b703-d1e31fa5b208/runs?additionalRuns[]=0e9dbb39-d5cb-45d8-8d84-97b1dc077838

**Pull request merged**
